### PR TITLE
The Python3 dev doc is more general than just modules

### PR DESCRIPTION
**Pull request by abadger**
Rename it to make that obvious.
##### ISSUE TYPE
 - Docs Pull Request

##### COMPONENT NAME
developing_python3

##### ANSIBLE VERSION

```
devel
```

### CODE REVIEW

**Comment by abadger** [comment]
Also added a section on formatting of strings.

**Comment by gundalow** [comment]
Looks good, though a few ideas 

1) Although bot developer focused, is https://docs.ansible.com/ansible/python_3_support.htmlbstill correct 

2) https://docs.ansible.com/ansible/python_3_support.htmlbstill should link to the developer pages

File: docs/docsite/rst/dev_guide/developing_python3.rst [hunk 1]
```diff
@@ -240,6 +254,92 @@ We do not prefix the text strings instead because we only operate
...
+    try:
+        a = 2/0
+    except ValueError as e:
+        module.fail_json(msg="Tried to divide by zero: %s" % e)
+
+Octal numbers
+-------------
+
+In Python-2.6, octal literals could be specified as ``0755``.  In Python-3, that is
+invalid and octals must be specified as ``0o755``.
```
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_python3.rst]: Please delete the words 'that is invalid and'.

File: docs/docsite/rst/dev_guide/developing_python3.rst [hunk 2]
```diff
@@ -240,6 +254,92 @@ We do not prefix the text strings instead because we only operate
...
+
+String formatting
+-----------------
+
+Python-2.6 has less featureful .format()
```
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_python3.rst]: Please replace the non-word 'featureful' with 'functional'.

File: docs/docsite/rst/dev_guide/developing_python3.rst [hunk 3]
```diff
@@ -240,6 +254,92 @@ We do not prefix the text strings instead because we only operate
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Starting in Python-2.6, strings gained a method called ``format()`` to put
+strings together.  However, one commonly used
+feature of ``format()`` wasn't added until Python-2.7 so you need to remember
```
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_python3.rst]: comma after 'Python-2.7'

**Review by dharmabumstead — CHANGES_REQUESTED** [review]
Looks pretty good - only a few changes. Thanks!

**Comment by gundalow** [comment]
3) Under "Porting Code"
Maybe reword as supporting Python 2 and Python 3
Add a line regarding this isn't about doing a one time port to only work on Python 3

File: docs/docsite/rst/dev_guide/developing_python3.rst [hunk 4]
```diff
@@ -274,67 +374,42 @@ coded to expect bytes on Python-2 and text on Python-3.
 Tips, tricks, and idioms to adopt
 =================================
 
-Exceptions
-----------
+Old Exception Syntax
```
Comment (gundalow) [inline docs/docsite/rst/dev_guide/developing_python3.rst]: "old" is a relative term, I think a more precise title using versions would be better.

File: docs/docsite/rst/dev_guide/developing_python3.rst [hunk 5]
```diff
@@ -274,67 +374,42 @@ coded to expect bytes on Python-2 and text on Python-3.
...
+Python-2 and Python-3.  You may still see this used in some modules:
+
+.. code-block:: python
+
+    from ansible.module_utils.pycompat24 import get_exception
+    [...]
 
     try:
         a = 2/0
-    except ValueError as e:
```
Comment (gundalow) [inline docs/docsite/rst/dev_guide/developing_python3.rst:285]: I wonder if something has been lost here, maybe this could be expanded to

A) This is how you should detail with exceptions in a way that works with Python 2.6+ & Python 3.5+
Code sample and explain why this is the correct way

B) Example of Python-2.4 only bad code

C) Example of python 2.4+ & Python 3.5+ exception handling
Note that this is only for Ansible 2.? - 2.3

D) Guidelines regarding if/when you should refactoring code from (B) or (C) to(A)

File: docs/docsite/rst/dev_guide/developing_python3.rst (hunk 2 above)
Comment (abadger) [reply docs/docsite/rst/dev_guide/developing_python3.rst]: What would be a word that means it doesn't have a feature that is in later versions?  It still performs the same functions but only by taking differently formatted input.

File: docs/docsite/rst/dev_guide/developing_python3.rst (hunk 5 above)
Comment (abadger) [reply docs/docsite/rst/dev_guide/developing_python3.rst:285]: (A) is what is there now.  Added (B).    (C) is documented in the module section. (D) is a global thing and we haven't really nailed down a decision here.  I skirt around the issue i nthe module section with this wording:

"Unless a change is going to be backported to Ansible-2.3, you should not
have to use this in new code." (where "this" refers to the get_exception() based coding style).

**Review by gundalow — APPROVED** [review]

File: docs/docsite/rst/python_3_support.rst [hunk 6]
```diff
@@ -15,26 +15,56 @@ Testing Python 3 with commands and playbooks
...
+.. code-block:: shell
+
+  ansible localhost -m ping -e 'ansible_python_interpreter=/usr/bin/python3'
+  ansible-playbook sample-playbook.yml -e 'ansible_python_interpreter=/usr/bin/python3'
+
+What to do if an incompatibility is found
+-----------------------------------------
+
+If you find a bug while testing modules with Python3 you can submit a bug
+report on `Ansible's GitHUb project
```
Comment (gundalow) [inline docs/docsite/rst/python_3_support.rst:64]: GitHUb -> GitHub

Also is the formatting for the URL correct, the rendering doesn't look right

Comment (gundalow) [reply docs/docsite/rst/python_3_support.rst:64]: Fixed in https://github.com/ansible/ansible/commit/01e27b108338c660446560c278aa2abfe76d556a
